appmap/pytest.py

```diff
@@ -0,0 +1,129 @@
+import importlib.metadata
+import logging
+import pytest
+import os
+import os.path
+import re
+
+import inflection
+
+from appmap._implementation import configuration, env, generation, recording
+from appmap._implementation.utils import fqname
+
+logger = logging.getLogger(__name__)
+
+
+class FuncItem:
+    def __init__(self, pyfuncitem):
+        self.item = pyfuncitem
+
+    @property
+    def class_name(self):
+        return self.item.cls.__name__
+
+    @property
+    def defined_class(self):
+        return fqname(self.item.cls)
+
+    @property
+    def method_id(self):
+        return self.item.originalname
+
+    def has_feature_group(self):
+        return self.item.cls is not None
+
+    @property
+    def feature_group(self):
+        return inflection.humanize(
+            inflection.titleize(self.class_name.removeprefix('Test'))
+        )
+
+    @property
+    def feature(self):
+        return inflection.humanize(self.test_name)
+
+    @property
+    def scenario_name(self):
+        name = f'{self.feature[0].lower()}{self.feature[1:]}'
+        if self.has_feature_group():
+            name = ' '.join([self.feature_group, name])
+        return name
+
+    @property
+    def test_name(self):
+        ret = self.item.name
+        return re.sub('^test.?_', '', ret)
+
+    @property
+    def filename(self):
+        fname = self.item.name
+        if self.item.cls:
+            fname = f'{self.defined_class}_{fname}'
+        fname = re.sub('[^a-zA-Z0-9-_]', '_', fname)
+        return fname.removesuffix('_')
+
+    @property
+    def metadata(self):
+        ret = {}
+        if self.has_feature_group():
+            ret['feature_group'] = self.feature_group
+            ret['recording'] = {
+                'defined_class': self.defined_class,
+                'method_id': self.method_id
```

Comment:
Won't these be present even without a feature group?

Reply:
Oops, yes, thanks. Looks like it got auto-indented, I'll fix it.

Reply:
Turns out this wasn't a bug, just a poorly named predicate.

There will only be a `defined_class` and `method_id` for this item if the test function is contained in class, which is what the `has_feature_group` predicate tested.

I've changed the name of the predicate so the code will be more clear.